README.md

decent
======

[![npm version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![coverage status][coveralls-image]][coveralls-url]
[![dependency status][david-image]][david-url]

`decent` is a decent Redis job queue for Node.js


# Motivation

There are powerful job queue modules for node.js + redis out there, like [kue](https://github.com/LearnBoost/kue) and [bull](https://github.com/OptimalBits/bull), to name a few. But powerful API comes at a price, they need complex data structure and redis scripts to achieve features such as delayed job, pause/resume and full text search. And since redis doesn't have traditional transaction, ie. [no rollback when one of the command failed](http://redis.io/topics/transactions), and [doesn't trigger error in node-redis driver](https://github.com/mranney/node_redis/issues/689), things can go south without developers noticing. Plus it's difficult to figure out what really happened due to non-intuitive redis data structure.

To us, the proper answer is to design around this problem, instead of adding more features, we want a job queue that's barebone, fully tested, easy to inspect, and doesn't hide errors from developers.


# Features

- Simple API with promise, works with your co/koa/whatever generator library.
- Automatic job recovery that ease queue shutdown and restart.
- Make use of native promise, and allow your favorite alternative.
- Only dependency is the redis driver.


# Install

`npm install decent --save`


# Usage

TODO


# API

## decent(name, opts)

Create a queue with `name` and config redis client connection based on `opts`, returns a decent queue instance.

### examples

```
var decent = require('decent');

var queue1 = decent('q1');
var queue2 = decent('q2', { 
	port: 6379
	, host: 'localhost'
	, connect_timeout: 5000 
});
```

### opts

- `port`: redis server port, default to `6379`
- `host`: redis server host, default to `'127.0.0.1'`
- `blockTimeout`: how long a client should wait for next job (see redis document on blocking command, such as [BLPOP](http://redis.io/commands/BLPOP)), defaults to `60` seconds, `0` to block forever.
- `maxRetry`: how many retries a job can have before being moved to failure queue, defaults to `3`, `0` to disable retry.
- and all [redis client options](https://github.com/mranney/node_redis#rediscreateclient).


## queue.add(data, opts)

Create a job on queue using `data` as payload and allows job specific `opts`, returns a promise that resolve to the created job.

### examples

```
queue.add({ a: 1 }).then(function(job) {
	console.log(job.data); // { a: 1 }
});

queue.add({ a: 1, b: 1 }, { retry: 1, timeout: 120 }).then(function(job) {
	console.log(job.data); // { a: 1, b: 1 }
	console.log(job.retry); // 1
	console.log(job.timeout); // 120
});
```

### opts

- `retry`: set initial retry counter, default to `0`
- `timeout`: set worker timeout in seconds, default to `60`

### job

- `id`: job id
- `data`: payload
- `retry`: current retry count for this job
- `timeout`: how many seconds a worker can run before it's terminated.
- `queue`: which queue this job currently belongs to.


## queue.worker(handler)

Register a handler function that process jobs, and start processing jobs in queue.

### examples

```
queue.worker(function(job, done) {
	setTimeout(function() {
		console.log(job.data);
		done();
	}, 100);
});
```

### done(err);

Must be called to signal the completion of job processing.

If called with an instance of `Error`, then `decent` will assume worker failed to process this job.

Fail jobs are moved back to work queue when they are below retry threshold, otherwise they are moved to failure queue.


## queue.count(name)

Returns a promise that resolve to the queue length of specified queue, default to `work` queue.

### examples

```
queue.count('work').then(function(count) {
	console.log(count); // pending job count
});

queue.count('run').then(function(count) {
	console.log(count); // running job count
});

queue.count('fail').then(function(count) {
	console.log(count); // failed job count
});
```


## queue.get(id)

Returns a promise that resolve to the job itself.

### examples

```
queue.get(1).then(function(job) {
	console.log(job.id); // 1
});
```


## queue.remove(id, name)

Returns a promise that will resolve when job is removed from redis (both job data and job queue). Default queue is `work`.

Note: `remove` does not return the job, use `get` then `remove` instead.

### examples

```
queue.remove(1, 'run').then(function() {
	// job has been removed from redis
});
```


## queue.stop()

Instructs queue worker to terminate gracefully on next loop. See events on how to monitor queue.

### examples

```
// ... setup queue and worker

queue.on('queue stop', function() {
	console.log('queue stopped gracefully');
});
queue.stop();
```


## queue.restart()

Restarts the queue worker loop. See events on how to monitor queue.

### examples

```
// ... setup queue and worker

queue.on('queue start', function() {
	console.log('queue restarted');
});
queue.restart();
```


# Events

`decent` is an instance of `EventEmitter`, so you can use `queue.on('event', func)` as usual.

## queue worker related

- `queue.emit('queue start')`: queue loop has started.
- `queue.emit('queue work', job)`: queue worker begin to process a `job`.
- `queue.emit('queue ok', job)`: queue worker has processed a `job`.
- `queue.emit('queue error', err, job)`: queue worker has failed to processed a `job` and thrown `err`, will retry later.
- `queue.emit('queue failure', err, job)`: queue worker has failed to processed a `job` and thrown `err`, retry limit reached.
- `queue.emit('queue exit', err)`: queue loop has terminated due to unhandled `err`.
- `queue.emit('queue stop')`: queue loop has stopped gracefully.


# License

MIT

[npm-image]: https://img.shields.io/npm/v/decent.svg?style=flat-square
[npm-url]: https://www.npmjs.com/package/decent
[travis-image]: https://img.shields.io/travis/bitinn/decent.svg?style=flat-square
[travis-url]: https://travis-ci.org/bitinn/decent
[coveralls-image]: https://img.shields.io/coveralls/bitinn/decent.svg?style=flat-square
[coveralls-url]: https://coveralls.io/r/bitinn/decent
[david-image]: https://img.shields.io/david/bitinn/decent.svg?style=flat-square
[david-url]: https://david-dm.org/bitinn/decent


CHANGELOG.md

Changelog
=========


# 1.x release

## v1.0.1 (master)

- Fix: node v0.10 promise support

## v1.0.0

- Feature: job item now includes their current queue name for better tracking
- Enhance: remove obsolete code and tests
- Break: Redis client and add/remove events are removed


# 0.x release

## v0.2.0

- Feature: new event `queue work` before worker start processing each job
- Feature: `add error` and `queue error` events now emits related job as second parameter
- Feature: `remove` can now remove job from any queue
- Fix: prevent potential memory leak with `run` loop due to unresolved promise chain

## v0.1.6

- Fix: stalled runQueue job are moved to workQueue on startup
- Enhance: verbose error message
- Enhance: reached 100% code coverage

## v0.1.5

- Fix: async worker api

## v0.1.4

- Fix: run loop

## v0.1.3

- initial public release



lib/dq.js

/**
 * dq.js
 *
 * the client and worker object
 */

'use strict';

var Redis = require('redis');
var Promise = require('native-or-bluebird');
var EventEmitter = require('events').EventEmitter;
var inherits = require('util').inherits;

inherits(Queue, EventEmitter);

module.exports = Queue;

/**
 * Create an instance of Queue
 *
 * @param   String  name  Name of this queue
 * @param   Object  opts  Redis options
 * @return  Object
 */
function Queue(name, opts) {

	if (!name) {
		throw new Error('queue name required');
	}

	opts = opts || {};

	// basic
	this.name = name;
	this.port = opts.port || 6379;
	this.host = opts.host || '127.0.0.1';

	// queue config
	this.config = {
		blockTimeout: opts.blockTimeout || 30
		, maxRetry: opts.maxRetry || 3
	};

	// redis options
	this.opts = opts;

	// main redis client
	this.client = Redis.createClient(this.port, this.host, this.opts);

	this.client.on('ready', this.clientReady.bind(this));
	this.client.on('connect', this.clientConnect.bind(this));
	this.client.on('error', this.clientError.bind(this));
	this.client.on('end', this.clientEnd.bind(this));
	this.client.on('drain', this.clientDrain.bind(this));
	this.client.on('idle', this.clientIdle.bind(this));

	// client for blocking command only
	this.bclient = Redis.createClient(this.port, this.host, this.opts);

	// queue name
	this.prefix = 'dq:' + this.name;
	this.workQueue = this.prefix + ':work';
	this.runQueue = this.prefix + ':run';
	this.failQueue = this.prefix + ':fail';

	// status code
	this.status_timeout = 1;

}

/**
 * Increment queue id and return it
 *
 * @return  Promise
 */
Queue.prototype.nextId = function() {

	var self = this;

	return new Promise(function(resolve, reject) {

		self.client.incr(self.prefix + ':id', function(err, res) {
			if (err) {
				reject(err);
			} else {
				resolve(res);
			}
		});

	});

};

/**
 * Retrieve the next job on work queue, put it on the run queue
 *
 * @return  Promise
 */
Queue.prototype.nextJob = function() {

	var self = this;

	return new Promise(function(resolve, reject) {

		self.bclient.brpoplpush(
			self.workQueue
			, self.runQueue
			, self.config.blockTimeout
			, function(err, id) {
				// client error
				if (err) {
					reject(err);

				// blocking timeout, return special code
				} else if (id === null) {
					resolve(self.status_timeout);

				} else {
					resolve(self.get(id));
				}
			}
		);

	});

};

/**
 * Convert job data into a format supported by redis client
 *
 * @param   Object  job  Queue format
 * @return  Object
 */
Queue.prototype.toClient = function(job) {

	// all values must be primitive type
	return {
		id: job.id
		, data: JSON.stringify(job.data)
		, retry: job.retry
		, timeout: job.timeout
	};

};

/**
 * Convert redis data into the original job format
 *
 * @param   Object  job  Redis format
 * @return  Object
 */
Queue.prototype.fromClient = function(job) {

	// values in redis are stored as string
	return {
		id: parseInt(job.id, 10)
		, data: JSON.parse(job.data)
		, retry: parseInt(job.retry, 10)
		, timeout: parseInt(job.timeout, 10)
	};

};

/**
 * Register a handler function that process each job
 *
 * @param   Function  handler  Process job data
 * @return  Void
 */
Queue.prototype.worker = function(handler) {

	if (typeof handler !== 'function') {
		throw new Error('worker must be a function');
	}

	if (this.handler) {
		throw new Error('worker can only be registered once');
	}

	this.handler = handler;

	// once the handler is registered, we can start processing jobs
	this.listen();

};

/**
 * Start listening for jobs on queue, handle worker function error
 *
 * @return  Promise
 */
Queue.prototype.listen = function() {

	var self = this;

	return this.run().catch(function(err) {
		self.emit('queue exit', err);
	});

};

/**
 * Repeatedly retrieve new job from work queue then process it
 *
 * @return  Promise
 */
Queue.prototype.run = function() {

	var self = this;

	return this.nextJob()
		.then(function(res) {
			return self.handleStatus(res);
		})
		.then(function(job) {
			return self.handleJob(job);
		})
		.then(function() {
			if (self.shutdown) {
				self.shutdown = false;
				self.emit('queue stop');
			} else {
				return self.run();
			}
		});

};

/**
 * Handle known status code from nextJob
 *
 * @param   Mixed  res  Result from nextJob
 * @return  Mixed
 */
Queue.prototype.handleStatus = function(res) {

	// blocking timeout, try again
	if (res === this.status_timeout) {
		return this.nextJob();

	// pass on result
	} else {
		return res;
	}

};

/**
 * Handle job from nextJob
 *
 * @param   Object   job  Formatted job
 * @return  Promise
 */
Queue.prototype.handleJob = function(job) {

	if (!this.handler) {
		throw new Error('worker must be registered first');
	}

	var self = this;

	try {
		this.handler(job);
	} catch(err) {
		// move job to next queue
		return this.moveJob(job).then(function() {
			// emit handler related error
			self.emit('queue error', err);
		});
	}

	// job done, remove it
	return new Promise(function(resolve, reject) {

		self.client.multi()
			.lrem(self.runQueue, 1, job.id)
			.del(self.prefix + ':' + job.id)
			.exec(function(err, res) {
				// client error
				if (err) {
					// only return the first error
					if (err.length > 0) {
						err = err[0];
					}

					reject(err);

				// see Queue.prototype.add comments on why these check are necessary

				} else if (res[0] != 1) {
					err = new Error('unable to remove job id');
					reject(err);
				} else if (res[1] != 1) {
					err = new Error('unable to remove job data');
					reject(err);
				} else {
					// job done event
					self.emit('queue ok', job);
					resolve();
				}
			});

	});

};

/**
 * Move job from run queue to another queue
 *
 * @param   Object   job  Formatted job
 * @return  Promise
 */
Queue.prototype.moveJob = function(job) {

	var self = this;

	// job done, remove it
	return new Promise(function(resolve, reject) {

		var multi = self.client.multi();

		// check retry limit
		if (job.retry >= self.config.maxRetry) {
			multi.rpoplpush(self.runQueue, self.failQueue);
		} else {
			multi.rpoplpush(self.runQueue, self.workQueue);
		}

		// update job retry count
		job.retry++;
		multi.hset(self.prefix + ':' + job.id, 'retry', job.retry);

		multi.exec(function(err, res) {
			// client error
			if (err) {
				// only return the first error
				if (err.length > 0) {
					err = err[0];
				}

				reject(err);

			// see Queue.prototype.add comments on why these check are necessary

			} else if (res[0] === null) {
				err = new Error('unable to find job id on queue');
				reject(err);
			// response should be 0
			} else if (isNaN(parseInt(res[1], '10')) || res[1] > 0) {
				err = new Error('unable to update job retry count');
				reject(err);
			} else {
				resolve();
			}
		});

	});

};

/**
 * Stop queue processing gracefully
 *
 * @return  Void
 */
Queue.prototype.stop = function() {

	this.shutdown = true;

};

/**
 * Restart queue processing
 *
 * @return  Promise
 */
Queue.prototype.restart = function() {

	// TODO

};

/**
 * Re-run all previously failed jobs
 *
 * @return  Promise
 */
Queue.prototype.runFailJobs = function() {

	// TODO

};

/**
 * Report the current number of jobs in work queue
 *
 * @param   String   name  Queue name
 * @return  Promise
 */
Queue.prototype.count = function(name) {

	name = name || 'work';
	var self = this;

	return new Promise(function(resolve, reject) {

		if (!self[name + 'Queue']) {
			reject(new Error('invalid queue name'));
			return;
		}

		self.client.llen(self[name + 'Queue'], function(err, res) {
			if (err) {
				reject(err);
			} else {
				resolve(res);
			}
		});

	});

};

/**
 * Return a job without removing it from redis
 *
 * @param   Number   id  Job id
 * @return  Promise
 */
Queue.prototype.get = function(id) {

	var self = this;

	return new Promise(function(resolve, reject) {

		if (!id) {
			reject(new Error('job id missing'));
			return;
		}

		// get job
		self.client.hgetall(self.prefix + ':' + id, function(err, job) {
			// client error
			if (err) {
				reject(err);

			// job data missing
			} else if (job === null) {
				reject(new Error('job data missing'));

			} else {
				// format job for client, handle invalid job
				try {
					job = self.fromClient(job);
				} catch(err) {
					reject(err);
				}
				
				resolve(job);
			}
		});

	});

};

/**
 * Add a job onto the work queue, overwrite duplicate job
 *
 * @param   Object   data  Actual data for worker to process
 * @param   Object   opts  Additional job options
 * @return  Promise
 */
Queue.prototype.add = function(data, opts) {

	data = data || {};
	opts = opts || {};

	var self = this;

	// note that queue id always increment, even if we don't use it
	return this.nextId().then(function(id) {

		// job structure
		var job = {
			id: opts.id || id
			, data: data
			, retry: opts.retry || 0
			, timeout: opts.timeout || 60
		};

		// format job for redis, invalid data will reject promise
		var rjob = self.toClient(job);

		// add job as hash, push its id onto queue
		// note that overwrite existing job will requeue job id
		return new Promise(function(resolve, reject) {

			self.client.multi()
				.hmset(self.prefix + ':' + rjob.id, rjob)
				.lrem(self.workQueue, 1, rjob.id)
				.lpush(self.workQueue, rjob.id)
				.exec(function(err, res) {
					// client error
					if (err) {
						// only return the first error
						if (err.length > 0) {
							err = err[0];
						}

						self.emit('add error', err);
						reject(err);

					// command failure
					// we need to check each command is returning expected result
					// as err is null in this case, ref: http://git.io/bT5C4Q

					} else if (res[0] !== 'OK') {
						err = new Error('unable to set job item');
						self.emit('add error', err);
						reject(err);
					// response should be 0 or 1
					} else if (isNaN(parseInt(res[1], '10')) || res[1] > 1) {
						err = new Error('unable to purge duplicate job id');
						self.emit('add error', err);
						reject(err);
					// response should be a number and at least 1
					} else if (isNaN(parseInt(res[2], '10')) || res[2] < 1) {
						err = new Error('unable to add job id');
						self.emit('add error', err);
						reject(err);
					} else {
						self.emit('add ok', job);
						resolve(job);
					}
				});

		});

	})

};

/**
 * Handle redis client ready event
 *
 * @return  Void
 */
Queue.prototype.clientReady = function() {

	if (!this.opts.no_ready_check) {
		this.emit('client ready');
	}

};

/**
 * Handle redis client connect event
 *
 * @return  Void
 */
Queue.prototype.clientConnect = function() {

	if (this.opts.no_ready_check) {
		this.emit('client ready');
	}

};

/**
 * Handle redis client error event
 *
 * @param   Object  err  Error from redis client
 * @return  Void
 */
Queue.prototype.clientError = function(err) {

	if (err instanceof Error) {
		this.emit('client error', err);
	}

};

/**
 * Handle redis client end event
 *
 * @return  Void
 */
Queue.prototype.clientEnd = function() {

	this.emit('client close');

};

/**
 * Handle redis client drain event
 *
 * @return  Void
 */
Queue.prototype.clientDrain = function() {

	var l = this.client.command_queue.length;
	if (l > 0) {
		this.idle = false;
		this.emit('client pressure', l);
	}

};

/**
 * Handle redis client idle event
 *
 * @return  Void
 */
Queue.prototype.clientIdle = function() {

	if (!this.idle) {
		this.idle = true;
		this.emit('client pressure', 0);
	}

};



lib/decent.js

/**
 * decent.js
 *
 * the client and worker object
 */

'use strict';

var Redis = require('redis');
var EventEmitter = require('events').EventEmitter;
var inherits = require('util').inherits;

inherits(Queue, EventEmitter);

module.exports = Queue;

/**
 * Create an instance of Queue
 *
 * @param   String  name  Name of this queue
 * @param   Object  opts  Redis options
 * @return  Object
 */
function Queue(name, opts) {

	// allow call as function
	if (!(this instanceof Queue))
		return new Queue(name, opts);

	if (!name) {
		throw new Error('queue name is required');
	}

	opts = opts || {};

	// basic
	this.name = name;
	this.port = opts.port || 6379;
	this.host = opts.host || '127.0.0.1';

	// queue config
	this.config = {
		blockTimeout: opts.blockTimeout || 60
		, maxRetry: opts.maxRetry || 3
	};

	// redis options
	this.opts = opts;

	// main redis client
	this.client = Redis.createClient(this.port, this.host, this.opts);

	// client for blocking command only
	this.bclient = Redis.createClient(this.port, this.host, this.opts);

	// queue name
	this.prefix = 'dq:' + this.name;
	this.workQueue = this.prefix + ':work';
	this.runQueue = this.prefix + ':run';
	this.failQueue = this.prefix + ':fail';

	// status code
	this.status_timeout = 1;

	// queue status
	this.shutdown = false;
	this.running = false;

}

/**
 * Add a job onto the work queue, overwrite duplicate job
 *
 * @param   Object   data  Actual data for worker to process
 * @param   Object   opts  Additional job options
 * @return  Promise
 * @api     Public
 */
Queue.prototype.add = function(data, opts) {

	var self = this;

	// note that queue id always increment, even if we don't use it
	return this.nextId().then(function(id) {

		if (!data || typeof data !== 'object') {
			throw new Error('job data payload must be an object');
		}

		opts = opts || {};

		// job structure
		var job = {
			id: opts.id || id
			, data: data
			, retry: opts.retry || 0
			, timeout: opts.timeout || 60
			, queue: self.workQueue
		};

		// format job for redis, invalid data will reject promise
		var rjob = self.toClient(job);

		// add job as hash, push its id onto queue
		// note that overwrite existing job will requeue job id
		return new Queue.Promise(function(resolve, reject) {

			self.client.multi()
				.hmset(self.prefix + ':' + rjob.id, rjob)
				.lrem(self.workQueue, 1, rjob.id)
				.lpush(self.workQueue, rjob.id)
				.exec(function(err, res) {
					// client error
					if (err) {
						reject(err[0]);

					// command failure
					// we need to check commands are returning expected result
					// as err is null in this case, ref: http://git.io/bT5C4Q

					// duplicate job id should be purged
					} else if (isNaN(parseInt(res[1], '10')) || res[1] > 1) {
						err = new Error('fail to remove duplicate job id from queue');
						reject(err);
					} else {
						resolve(job);
					}
				});

		});

	})

};

/**
 * Remove a job from queue given the job id
 *
 * @param   Number   id    Job id
 * @param   String   name  Queue name
 * @return  Promise
 * @api     Public
 */
Queue.prototype.remove = function(id, name) {

	name = name || 'work';
	var self = this;

	// job done, remove it
	return new Queue.Promise(function(resolve, reject) {

		if (!self[name + 'Queue']) {
			reject(new Error('invalid queue name'));
			return;
		}

		self.client.multi()
			.lrem(self[name + 'Queue'], 1, id)
			.del(self.prefix + ':' + id)
			.exec(function(err, res) {
				// see prototype.add on why verbose checks are needed
				// client error
				if (err) {
					// only return the first error
					reject(err[0]);

				// command error
				} else if (res[0] != 1) {
					err = new Error('job id missing from queue');
					reject(err);
				} else if (res[1] != 1) {
					err = new Error('job data missing from redis');
					reject(err);
				} else {
					resolve();
				}
			});

	});

};

/**
 * Report the current number of jobs in work queue
 *
 * @param   String   name  Queue name
 * @return  Promise
 * @api     Public
 */
Queue.prototype.count = function(name) {

	name = name || 'work';
	var self = this;

	return new Queue.Promise(function(resolve, reject) {

		if (!self[name + 'Queue']) {
			reject(new Error('invalid queue name'));
			return;
		}

		self.client.llen(self[name + 'Queue'], function(err, res) {
			if (err) {
				reject(err);
			} else {
				resolve(res);
			}
		});

	});

};

/**
 * Return job data without removing it from redis
 *
 * @param   Number   id  Job id
 * @return  Promise
 * @api     Public
 */
Queue.prototype.get = function(id) {

	var self = this;

	return new Queue.Promise(function(resolve, reject) {

		if (!id) {
			reject(new Error('job id required'));
			return;
		}

		// get job
		self.client.hgetall(self.prefix + ':' + id, function(err, job) {
			// client error
			if (err) {
				reject(err);

			} else if (job === null) {
				reject(new Error('job data missing from redis'));

			} else {
				// format job for client, handle invalid job
				try {
					job = self.fromClient(job);
				} catch(err) {
					reject(err);
				}
				
				resolve(job);
			}
		});

	});

};

/**
 * Stop queue processing gracefully
 *
 * @return  Void
 * @api     Public
 */
Queue.prototype.stop = function() {

	this.shutdown = true;

};

/**
 * Restart queue processing
 *
 * @return  Void
 * @api     Public
 */
Queue.prototype.restart = function() {

	// prevent duplicate worker
	if (this.running) {
		throw new Error('worker is already running');
	}

	this.start();

};

/**
 * Register a handler function that process each job
 *
 * @param   Function  handler  Process job data
 * @return  Void
 * @api     Public
 */
Queue.prototype.worker = function(handler) {

	if (typeof handler !== 'function') {
		throw new Error('job handler must be a function');
	}

	if (this.handler) {
		throw new Error('job handler can only be registered once');
	}

	this.handler = handler;

	// once the handler is registered, we can start processing jobs
	this.start();

};

/**
 * Start listening for jobs on queue, handle worker function error
 *
 * @return  Void
 * @api     Private
 */
Queue.prototype.start = function() {

	var self = this;

	this.running = true;
	this.emit('queue start');

	this.run();

};

/**
 * Repeatedly retrieve new job from work queue then process it
 *
 * @return  Void
 * @api     Private
 */
Queue.prototype.run = function() {

	var self = this;

	// loop
	this.recoverJob()
		.then(function(res) {
			return self.readJob(res);
		})
		.then(function(job) {
			return self.handleJob(job);
		})
		.then(function() {
			// handle graceful shutdown
			if (self.shutdown) {
				self.shutdown = false;
				self.running = false;
				self.emit('queue stop');
				return;
			}
			// tail recursion
			self.run();
		}, function(err) {
			// exit queue
			self.running = false;
			self.emit('queue exit', err);
		});

};

/**
 * Wait for job on work queue, loop until found
 *
 * @param   Mixed    res  Can be empty, status code or job
 * @return  Promise
 * @api     Private
 */
Queue.prototype.readJob = function(res) {

	var self = this;

	// first run, or blocking timeout
	if (!res || res === this.status_timeout) {
		return this.nextJob().then(function(res) {
			return self.readJob(res);
		});

	// pass on job object
	} else {
		return Queue.Promise.resolve(res);
	}

};

/**
 * Process job with handler
 *
 * @param   Object   job  Formatted job
 * @return  Promise
 * @api     Private
 */
Queue.prototype.handleJob = function(job) {

	var self = this;

	// this promise always resolve, errors are handled
	return new Queue.Promise(function(resolve, reject) {

		// start working on job
		self.emit('queue work', job);

		if (!self.handler) {
			reject(new Error('job handler must be registered first'));
			return;
		}

		// support job timeout
		if (job.timeout > 0) {
			setTimeout(function() {
				reject(new Error('job timeout threshold reached'));
			}, job.timeout * 1000);
		}

		// callback function for async worker
		var done = function(input) {
			if (input instanceof Error) {
				reject(input);
			} else {
				resolve();
			}
		};

		// catch worker error
		try {
			self.handler(job, done);
		} catch(err) {
			reject(err);
		}

	}).then(function() {

		// job done, remove it and emit event
		return self.remove(job.id, 'run').then(function() {
			self.emit('queue ok', job);
		});

	}).catch(function(err) {

		// job failure, move job to appropriate queue
		return self.moveJob(job).then(function(job) {
			if (job.queue === self.failQueue) {
				self.emit('queue failure', err, job);
			} else {
				self.emit('queue error', err, job);
			}
		});

	});

};

/**
 * Move job from run queue to another queue
 *
 * @param   Object   job  Formatted job
 * @return  Promise
 * @api     Private
 */
Queue.prototype.moveJob = function(job) {

	var self = this;

	return new Queue.Promise(function(resolve, reject) {

		var multi = self.client.multi();

		// check retry limit, decide next queue
		if (job.retry >= self.config.maxRetry) {
			multi.rpoplpush(self.runQueue, self.failQueue);
			job.queue = self.failQueue;
		} else {
			multi.rpoplpush(self.runQueue, self.workQueue);
			job.queue = self.workQueue;
		}

		// update job retry count
		job.retry++;
		multi.hset(self.prefix + ':' + job.id, 'retry', job.retry);
		multi.hset(self.prefix + ':' + job.id, 'queue', job.queue);

		multi.exec(function(err, res) {
			// see prototype.add on why verbose checks are needed
			// client error
			if (err) {
				// only return the first error
				reject(err[0]);

			// command error
			} else if (res[0] === null) {
				err = new Error('job id missing from queue');
				reject(err);
			} else if (isNaN(parseInt(res[1], '10')) || res[1] > 0) {
				err = new Error('partial job data, retry count missing');
				reject(err);
			} else if (isNaN(parseInt(res[2], '10')) || res[2] > 0) {
				err = new Error('partial job data, queue name missing');
				reject(err);
			} else {
				resolve(job);
			}
		});

	});

};

/**
 * Recover job from runQueue back to start of workQueue
 *
 * @return  Promise
 * @api     Private
 */
Queue.prototype.recoverJob = function() {

	var self = this;

	return this.count('run').then(function(count) {

		// nothing to do if last session shutdown gracefully
		if (count === 0) {
			return;

		// by design there should only be at max 1 job in runQueue
		} else if (count > 1) {
			return Queue.Promise.reject(
				new Error('more than 1 job in queue, purge manually')
			);

		// move the job to work queue, note that retry count does not increase
		} else {
			return new Queue.Promise(function(resolve, reject) {
				self.client.rpoplpush(self.runQueue, self.workQueue, function(err, res) {
					if (err) {
						reject(err);
					} else if (res === null) {
						reject(new Error('job id missing from queue')); 
					} else {
						resolve();
					}
				});
			});
		}

	});

};

/**
 * Increment queue id and return it
 *
 * @return  Promise
 * @api     Private
 */
Queue.prototype.nextId = function() {

	var self = this;

	return new Queue.Promise(function(resolve, reject) {

		self.client.incr(self.prefix + ':id', function(err, res) {
			if (err) {
				reject(err);
			} else {
				resolve(res);
			}
		});

	});

};

/**
 * Retrieve the next job on work queue, put it on the run queue
 *
 * @return  Promise
 * @api     Private
 */
Queue.prototype.nextJob = function() {

	var self = this;

	return new Queue.Promise(function(resolve, reject) {

		self.bclient.brpoplpush(
			self.workQueue
			, self.runQueue
			, self.config.blockTimeout
			, function(err, id) {
				// client error
				if (err) {
					reject(err);

				// blocking timeout, return special code
				} else if (id === null) {
					resolve(self.status_timeout);

				} else {
					resolve(self.get(id));
				}
			}
		);

	});

};

/**
 * Convert job data into a format supported by redis client
 *
 * @param   Object   job  Queue format
 * @return  Object
 * @api     Private
 */
Queue.prototype.toClient = function(job) {

	// all values must be primitive type
	return {
		id: job.id
		, data: JSON.stringify(job.data)
		, retry: job.retry
		, timeout: job.timeout
		, queue: job.queue
	};

};

/**
 * Convert redis data into the original job format
 *
 * @param   Object   job  Redis format
 * @return  Object
 * @api     Private
 */
Queue.prototype.fromClient = function(job) {

	// values in redis are stored as string
	return {
		id: parseInt(job.id, 10)
		, data: JSON.parse(job.data)
		, retry: parseInt(job.retry, 10)
		, timeout: parseInt(job.timeout, 10)
		, queue: job.queue
	};

};

// allow custom promise module
Queue.Promise = global.Promise;
